# Capstone 2 Project

## NLP training on ca. 3 million Yelp reviews

## What is covered in this part of the project
This part of the project deals with natural language processing of the Yelp data set. Specifically, it will import the raw data in json format, filter the reviews that relate to restaurants, prepare the data for NLP and perform a number of NLP functions:

1. Text preparation, including tokenization, lemmatization and normalization of the text using the spaCy library
1. Phrase modeling using gensim
1. Topic modeling with LDA
1. Word vector models with word2vec
1. Clustering words that commonly appear together 

Topic modeling and word2vec clusters will provide context to the reviews in order to return summary information that is easily interpretable and understandable to an average user, and may provide useful insight.

## Loading the Yelp data
[**The Yelp Dataset**](https://www.yelp.com/dataset_challenge/) is a dataset published by the business review service [Yelp](http://yelp.com) for academic research and educational purposes. I really like the Yelp dataset as a subject for machine learning and natural language processing demos, because it's big (but not so big that you need your own data center to process it), well-connected, and anyone can relate to it &mdash; it's largely about food, after all!

**Note:** If you'd like to execute this notebook interactively on your local machine, you'll need to download your own copy of the Yelp dataset. 

After filtering for restaurants, there are approximately __52K__ restaurants with approximately __2.9M__ user reviews related to them.

The raw data is available in six of files in _.json_ format, of which two are relevant for the project:
- __business.json__ &mdash; _the records for individual businesses_
- __review.json__ &mdash; _the records for reviews users wrote about businesses_

The files are text files (UTF-8) with one _json object_ per line, each one corresponding to an individual data record. 

In [1]:
import os
import codecs

data_directory = os.path.join('data',
                              'dataset')

businesses_filepath = os.path.join(data_directory,
                                   'business.json')

with codecs.open(businesses_filepath, encoding='utf_8') as f:
    first_business_record = f.readline() 

print first_business_record

{"business_id": "YDf95gJZaq05wvo7hTQbbQ", "name": "Richmond Town Square", "neighborhood": "", "address": "691 Richmond Rd", "city": "Richmond Heights", "state": "OH", "postal_code": "44143", "latitude": 41.5417162, "longitude": -81.4931165, "stars": 2.0, "review_count": 17, "is_open": 1, "attributes": {"RestaurantsPriceRange2": 2, "BusinessParking": {"garage": false, "street": false, "validated": false, "lot": true, "valet": false}, "BikeParking": true, "WheelchairAccessible": true}, "categories": ["Shopping", "Shopping Centers"], "hours": {"Monday": "10:00-21:00", "Tuesday": "10:00-21:00", "Friday": "10:00-21:00", "Wednesday": "10:00-21:00", "Thursday": "10:00-21:00", "Sunday": "11:00-18:00", "Saturday": "10:00-21:00"}}



The business records consist of _key, value_ pairs containing information about the particular business. The information from this file is treated in a separate notebook named _"Restaurants"_. 

This notebook will process information contained in the _reviews.json_ file.

In [2]:
review_json_filepath = os.path.join(data_directory,
                                    'review.json')

with codecs.open(review_json_filepath, encoding='utf_8') as f:
    first_review_record = f.readline()
    
print first_review_record

{"review_id":"VfBHSwC5Vz_pbFluy07i9Q","user_id":"cjpdDjZyprfyDG3RlkVG3w","business_id":"uYHaNptLzDLoV_JZ_MuzUA","stars":5,"date":"2016-07-12","text":"My girlfriend and I stayed here for 3 nights and loved it. The location of this hotel and very decent price makes this an amazing deal. When you walk out the front door Scott Monument and Princes street are right in front of you, Edinburgh Castle and the Royal Mile is a 2 minute walk via a close right around the corner, and there are so many hidden gems nearby including Calton Hill and the newly opened Arches that made this location incredible.\n\nThe hotel itself was also very nice with a reasonably priced bar, very considerate staff, and small but comfortable rooms with excellent bathrooms and showers. Only two minor complaints are no telephones in room for room service (not a huge deal for us) and no AC in the room, but they have huge windows which can be fully opened. The staff were incredible though, letting us borrow umbrellas for t

A few attributes of note on the review records:
- __business\_id__ &mdash; _identifies the business in question and provides a link to each business in the business.json file_
- __text__ &mdash; _the actual text of the review written by the user_


Work required to prepare the text for NLP analysis requires the following:
1. Loading each business record line by line using _json.loads_. Python converts json objects into a Python dict
2. Filter each business record to include only those that have "Restaurant: in the category list
3. Form a set of the business_id records in order to use this filter when processing the reviews file

In [3]:
import json

restaurant_ids = set()

# open the businesses file
with codecs.open(businesses_filepath, encoding='utf_8') as f:
    
    # iterate through each line (json record) in the file
    for business_json in f:
        
        # convert the json record to a Python dict
        business = json.loads(business_json)
        
        # if this business is not a restaurant, skip to the next one
        if u'Restaurants' not in business[u'categories']:
            continue
            
        # add the restaurant business id to our restaurant_ids set
        restaurant_ids.add(business[u'business_id'])

# turn restaurant_ids into a frozenset, as we don't need to change it anymore
restaurant_ids = frozenset(restaurant_ids)

# print the number of unique restaurant ids in the dataset
print '{:,}'.format(len(restaurant_ids)), u'restaurants in the dataset.'

51,613 restaurants in the dataset.


This step creates a new file that contains only the text from reviews about restaurants, with one review per line in the file.

In [4]:
intermediate_directory = os.path.join('data', 'intermediate')

review_txt_filepath = os.path.join(intermediate_directory,
                                   'review_text_all.txt')

In [5]:
%%time

# this part is complete so make the statement false to skip directly to reading the file
if 0 == 1:
    
    review_count = 0

    # create & open a new file in write mode
    with codecs.open(review_txt_filepath, 'w', encoding='utf_8') as review_txt_file:

        # open the existing review json file
        with codecs.open(review_json_filepath, encoding='utf_8') as review_json_file:

            # loop through all reviews in the existing file and convert to dict
            for review_json in review_json_file:
                review = json.loads(review_json)

                # if this review is not about a restaurant, skip to the next one
                if review[u'business_id'] not in restaurant_ids:
                    continue

                # write the restaurant review as a line in the new file
                # escape newline characters in the original review text
                review_txt_file.write(review[u'text'].replace('\n', '\\n') + '\n')
                review_count += 1

    print u'''Text from {:,} restaurant reviews
              written to the new txt file.'''.format(review_count)
    
else:
    
    with codecs.open(review_txt_filepath, encoding='utf_8') as review_txt_file:
        for review_count, line in enumerate(review_txt_file):
            pass
        
    print u'Text from {:,} restaurant reviews in the txt file.'.format(review_count + 1)

Text from 2,929,512 restaurant reviews in the txt file.
Wall time: 1min 26s


## Text preparation using spaCy

The [**spaCy**](https://spacy.io) library will be used to perform the following NLP tasks: 
- Tokenization
- Text normalization, including converting text to lemmatized form 
- Part-of-speech tagging
- Syntactic dependency parsing
- Sentence boundary detection
- Named entity recognition and annotation

spaCy provides downloadable English-language models that enable the comparison of a specific corpus of text to a general English language corpus, such as common stpowords, and the probability of occurence of words. 

In [6]:
import pandas as pd
import itertools as it

The next part is needed to append the system path list to access the location of the spaCy module and English language model on my laptop.

In [7]:
import sys
sys.path.append('C:\Windows\System32\.env\Lib\site-packages')

This part now handles the text processing using spaCy

In [8]:
import spacy
import en_core_web_sm
# English model will only load with this code, does not work using the code provided in spaCy documentation
nlp = en_core_web_sm.load() 

In [9]:
# this examines a sample review
with codecs.open(review_txt_filepath, encoding='utf_8') as f:
    sample_review = list(it.islice(f, 8, 9))[0]
    sample_review = sample_review.replace('\\n', '\n')
        
print sample_review

The staff here is great and they're nice,  wonderful and quick. People were ranting in raving about pei wei, I had to try it.  Even good yelp reviews.  I'm highly dissatisfied with the flavor of the food. This  should be labeled Asian inspired and not Asian. I've tried a variety of Chinese restaurants, this doesn't taste close to anything I've had at other Asian restaurants. Their Mongolian beef  was 5 pieces of beef and large mushrooms cut into thirds in a thick sauce. You eat the rice to wash off the nasty flavor. My shrimp was thickly coated in an overpowering  sauce as well.  I only ate some of the veggies that take center stage on a meat dish.  The center of my pork egg roll was cold. The hot N sour soup was a much thicker consistency almost like that of a chili instead of being brothy. Worst of all was the price.  This was not worth it to us. Neither me or my husband enjoyed either of  our dishes.  We didn't even eat half of our plates.  We even refused to take it home with us.  

Sentence detection and segmentation:

In [10]:
parsed_review = nlp(sample_review)

for num, sentence in enumerate(parsed_review.sents):
    print 'Sentence {}:'.format(num + 1)
    print sentence
    print ''

Sentence 1:
The staff here is great and they're nice,  wonderful and quick.

Sentence 2:
People were ranting in raving about pei wei, I had to try it.  

Sentence 3:
Even good yelp reviews.  

Sentence 4:
I'm highly dissatisfied with the flavor of the food.

Sentence 5:
This  should be labeled Asian inspired and not Asian.

Sentence 6:
I've tried a variety of Chinese restaurants, this doesn't taste close to anything I've had at other Asian restaurants.

Sentence 7:
Their Mongolian beef  was 5 pieces of beef and large mushrooms cut into thirds in a thick sauce.

Sentence 8:
You eat the rice to wash off the nasty flavor.

Sentence 9:
My shrimp was thickly coated in an overpowering  sauce as well.  

Sentence 10:
I only ate some of the veggies that take center stage on a meat dish.  

Sentence 11:
The center of my pork egg roll was cold.

Sentence 12:
The hot N sour soup was a much thicker consistency almost like that of a chili instead of being brothy.

Sentence 13:
Worst of all was the 

Named entity detection:

In [11]:
for num, entity in enumerate(parsed_review.ents):
    print 'Entity {}:'.format(num + 1), entity, '-', entity.label_
    print ''

Entity 1: Asian - NORP

Entity 2: Asian - NORP

Entity 3: Chinese - NORP

Entity 4: Asian - NORP

Entity 5: Mongolian - NORP

Entity 6: 5 - CARDINAL

Entity 7: half - CARDINAL

Entity 8: Asian - NORP



Part of speech tagging:

In [12]:
token_text = [token.orth_ for token in parsed_review]
token_pos = [token.pos_ for token in parsed_review]

pd.DataFrame(zip(token_text, token_pos),
             columns=['token_text', 'part_of_speech'])

token_text part_of_speech
0          The            DET
1        staff           NOUN
2         here            ADV
3           is           VERB
4        great            ADJ
5          and          CCONJ
6         they           PRON
7          're           VERB
8         nice            ADJ
9            ,          PUNCT
10                      SPACE
11   wonderful            ADJ
12         and          CCONJ
13       quick            ADJ
14           .          PUNCT
15      People           NOUN
16        were           VERB
17     ranting           VERB
18          in            ADP
19      raving           VERB
20       about            ADP
21         pei          PROPN
22         wei          PROPN
23           ,          PUNCT
24           I           PRON
25         had           VERB
26          to           PART
27         try           VERB
28          it           PRON
29           .          PUNCT
..         ...            ...
218    refused           VERB
219         to           PART
220       take           VERB
221         it           PRON
222       home            ADV
223       with            ADP
224         us           PRON
225          .          PUNCT
226                     SPACE
227         If            ADP
228        you           PRON
229       like           VERB
230        and          CCONJ
231      enjoy           VERB
232       what           NOUN
233    typical            ADJ
234      Asian            ADJ
235       food           NOUN
236     tastes           NOUN
237       like            ADP
238          ,          PUNCT
239                     SPACE
240         do           VERB
241        n't            ADV
242      waste           VERB
243       your            ADJ
244       time           NOUN
245       here            ADV
246          .          PUNCT
247         \n          SPACE

[248 rows x 2 columns]

Text normalization includes converting words to their lemmatized form so that grammatical form is removed in order to simplify and reduce the vocabulary. Hence words are converted to lowercase and words with the same stem but different grammatical forms are converted to a single lemma (e.g. "be:, "is", "were", "am", "are" are all represented as "be" in lemmatized form).

In [13]:
token_lemma = [token.lemma_ for token in parsed_review]
token_shape = [token.shape_ for token in parsed_review]

pd.DataFrame(zip(token_text, token_lemma, token_shape),
             columns=['token_text', 'token_lemma', 'token_shape'])

token_text token_lemma token_shape
0          The         the         Xxx
1        staff       staff        xxxx
2         here        here        xxxx
3           is          be          xx
4        great       great        xxxx
5          and         and         xxx
6         they      -PRON-        xxxx
7          're          be         'xx
8         nice        nice        xxxx
9            ,           ,           ,
10                                    
11   wonderful   wonderful        xxxx
12         and         and         xxx
13       quick       quick        xxxx
14           .           .           .
15      People      people       Xxxxx
16        were          be        xxxx
17     ranting        rant        xxxx
18          in          in          xx
19      raving        rave        xxxx
20       about       about        xxxx
21         pei         pei         xxx
22         wei         wei         xxx
23           ,           ,           ,
24           I      -PRON-           X
25         had        have         xxx
26          to          to          xx
27         try         try         xxx
28          it      -PRON-          xx
29           .           .           .
..         ...         ...         ...
218    refused      refuse        xxxx
219         to          to          xx
220       take        take        xxxx
221         it      -PRON-          xx
222       home        home        xxxx
223       with        with        xxxx
224         us      -PRON-          xx
225          .           .           .
226                                   
227         If          if          Xx
228        you      -PRON-         xxx
229       like        like        xxxx
230        and         and         xxx
231      enjoy       enjoy        xxxx
232       what        what        xxxx
233    typical     typical        xxxx
234      Asian       asian       Xxxxx
235       food        food        xxxx
236     tastes       taste        xxxx
237       like        like        xxxx
238          ,           ,           ,
239                                   
240         do          do          xx
241        n't         not         x'x
242      waste       waste        xxxx
243       your      -PRON-        xxxx
244       time        time        xxxx
245       here        here        xxxx
246          .           .           .
247         \n          \n          \n

[248 rows x 3 columns]

## Phrase Modeling with gensim

_Phrase modeling_ is an algorithm that passes over the text corpus to learn multi-word concepts. Using the algorithm once will detect phrases such as "ice cream" by detecting that "ice" and "cream" appear together with a frequency that passes a certain threshold. At the next pass, the algorithm will treat the bigram "ice_cream" as a single token and may also learn "vanilla_ice_cream" as a phrase and treat it as a trigram. 

This part makes three passes over the text corpus in order to reveal trigrams, which will help provide a robust level of meaning, given that the corpus is fairly large.

In [14]:
from gensim.models import Phrases
from gensim.models.word2vec import LineSentence

C:\Users\Work\Anaconda2\lib\site-packages\gensim\utils.py:865: UserWarning: detected Windows; aliasing chunkize to chunkize_serial
  warnings.warn("detected Windows; aliasing chunkize to chunkize_serial")
Using CNTK backend
C:\Users\Work\Anaconda2\lib\site-packages\keras\backend\cntk_backend.py:19: UserWarning: CNTK backend warning: GPU is not detected. CNTK's CPU version is not fully optimized,please run with GPU to get better performance.
  'CNTK backend warning: GPU is not detected. '


As we're performing phrase modeling, we'll be doing some iterative data transformation at the same time. Our roadmap for data preparation includes:

1. Segment text of complete reviews into sentences & normalize text
1. First-order phrase modeling $\rightarrow$ _apply first-order phrase model to transform sentences_
1. Second-order phrase modeling $\rightarrow$ _apply second-order phrase model to transform sentences_
1. Apply text normalization and second-order phrase model to text of complete reviews

We'll use this transformed data as the input for some higher-level modeling approaches in the following sections.

In [15]:
def punct_space(token):
    """
    helper function to eliminate tokens
    that are pure punctuation or whitespace
    """
    
    return token.is_punct or token.is_space

def line_review(filename):
    """
    generator function to read in reviews from the file
    and un-escape the original line breaks in the text
    """
    
    with codecs.open(filename, encoding='utf_8') as f:
        for review in f:
            yield review.replace('\\n', '\n')
            
def lemmatized_sentence_corpus(filename):
    """
    generator function to use spaCy to parse reviews,
    lemmatize the text, and yield sentences
    """
    
    for parsed_review in nlp.pipe(line_review(filename),
                                  batch_size=10000, n_threads=4):
        
        for sent in parsed_review.sents:
            yield u' '.join([token.lemma_ for token in sent
                             if not punct_space(token)])

In [16]:
unigram_sentences_filepath = os.path.join(intermediate_directory,
                                          'unigram_sentences_all.txt')

The `lemmatized_sentence_corpus` generator loops over the original review text, segmenting the reviews into individual sentences and normalizing the text. The output is written to a new file (`unigram_sentences_all`), with one normalized sentence per line.

In [17]:
%%time

# this is completed

if 0 == 1:

    with codecs.open(unigram_sentences_filepath, 'w', encoding='utf_8') as f:
        for sentence in lemmatized_sentence_corpus(review_txt_filepath):
            f.write(sentence + '\n')

Wall time: 0 ns


In [18]:
unigram_sentences = LineSentence(unigram_sentences_filepath)

The next phase passes over the text to link individual words into two-word phrases e.g. linking "`ice cream`" into a new, single token: "`ice_cream`".

In [19]:
bigram_model_filepath = os.path.join(intermediate_directory, 'bigram_model_all')

In [20]:
%%time

# this is completed

if 0 == 1:

    bigram_model = Phrases(unigram_sentences, max_vocab_size=10000000)

    bigram_model.save(bigram_model_filepath)
    
# load the finished model from disk
bigram_model = Phrases.load(bigram_model_filepath)

Wall time: 3.47 s


In [21]:
bigram_sentences_filepath = os.path.join(intermediate_directory,
                                         'bigram_sentences_all.txt')

In [22]:
%%time

# this is completed
# warning message: For a faster implementation, use the gensim.models.phrases.Phraser class
if 0 == 1:

    with codecs.open(bigram_sentences_filepath, 'w', encoding='utf_8') as f:
        
        for unigram_sentence in unigram_sentences:
            
            bigram_sentence = u' '.join(bigram_model[unigram_sentence])
            
            f.write(bigram_sentence + '\n')

Wall time: 0 ns


In [23]:
bigram_sentences = LineSentence(bigram_sentences_filepath)

In [24]:
trigram_model_filepath = os.path.join(intermediate_directory,
                                      'trigram_model_all')

In [25]:
%%time

# this is completed
if 0 == 1:

    trigram_model = Phrases(bigram_sentences, max_vocab_size=10000000)

    trigram_model.save(trigram_model_filepath)
    
# load the finished model from disk
trigram_model = Phrases.load(trigram_model_filepath)

Wall time: 4.98 s


In [26]:
trigram_sentences_filepath = os.path.join(intermediate_directory,
                                          'trigram_sentences_all.txt')

In [27]:
%%time

# this is completed
if 0 == 1:

    with codecs.open(trigram_sentences_filepath, 'w', encoding='utf_8') as f:
        
        for bigram_sentence in bigram_sentences:
            
            trigram_sentence = u' '.join(trigram_model[bigram_sentence])
            
            f.write(trigram_sentence + '\n')

Wall time: 0 ns


In [28]:
trigram_sentences = LineSentence(trigram_sentences_filepath)

The next step comprises the following: 
- running the complete text of the reviews through a pipeline that applies text normalization and phrase models
- removing stopwords
- writing the output to a new file with one review per line


In [29]:
trigram_reviews_filepath = os.path.join(intermediate_directory,
                                        'trigram_transformed_reviews_all.txt')

In [30]:
%%time

# this is completed
if 0 == 1:

    with codecs.open(trigram_reviews_filepath, 'w', encoding='utf_8') as f:
        
        for parsed_review in nlp.pipe(line_review(review_txt_filepath),
                                      batch_size=10000, n_threads=4):
            
            # lemmatize the text, removing punctuation and whitespace
            unigram_review = [token.lemma_ for token in parsed_review
                              if not punct_space(token)]
            
            # apply the first-order and second-order phrase models
            bigram_review = bigram_model[unigram_review]
            trigram_review = trigram_model[bigram_review]
            
            # remove any remaining stopwords
            trigram_review = [term for term in trigram_review
                              if term not in spacy.en.language_data.STOP_WORDS]
            
            # write the transformed review as a line in the new file
            trigram_review = u' '.join(trigram_review)
            f.write(trigram_review + '\n')

Wall time: 0 ns


## Topic Modeling with Latent Dirichlet Allocation (_LDA_)

In [31]:
from gensim.corpora import Dictionary, MmCorpus
from gensim.models.ldamulticore import LdaMulticore

import warnings
import cPickle as pickle

In [32]:
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.decomposition import NMF, LatentDirichletAllocation
from sklearn.datasets import fetch_20newsgroups

In [33]:
trigram_dictionary_filepath = os.path.join(intermediate_directory,
                                           'trigram_dict_all.dict')

In [34]:
%%time

# this is completed
if 0 == 1:

    trigram_reviews = LineSentence(trigram_reviews_filepath)

    # learn the dictionary by iterating over all of the reviews
    trigram_dictionary = Dictionary(trigram_reviews)
    
    # filter tokens that are very rare or too common from
    # the dictionary (filter_extremes) and reassign integer ids (compactify)
    trigram_dictionary.filter_extremes(no_below=10, no_above=0.4)
    trigram_dictionary.compactify()

    trigram_dictionary.save(trigram_dictionary_filepath)
    
# load the finished dictionary from disk
trigram_dictionary = Dictionary.load(trigram_dictionary_filepath)

Wall time: 69 ms


In [35]:
trigram_bow_filepath = os.path.join(intermediate_directory,
                                    'trigram_bow_corpus_all.mm')

In [36]:
def trigram_bow_generator(filepath):
    """
    generator function to read reviews from a file
    and yield a bag-of-words representation
    """
    
    for review in LineSentence(filepath):
        yield trigram_dictionary.doc2bow(review)

In [37]:
%%time

# this is completed
if 0 == 1:

    # generate bag-of-words representations for
    # all reviews and save them as a matrix
    MmCorpus.serialize(trigram_bow_filepath,
                       trigram_bow_generator(trigram_reviews_filepath))
    
# load the finished bag-of-words corpus from disk
trigram_bow_corpus = MmCorpus(trigram_bow_filepath)

Wall time: 971 ms


In [38]:
lda_model_filepath = os.path.join(intermediate_directory, 'lda_model_all')

In [39]:
%%time

# this is completed
if 0 == 1:

    with warnings.catch_warnings():
        warnings.simplefilter('ignore')
        
        # workers => sets the parallelism, and should be
        # set to your number of physical cores minus one
        lda = LdaMulticore(trigram_bow_corpus,
                           num_topics=50,
                           id2word=trigram_dictionary,
                           workers=3)
    
    lda.save(lda_model_filepath)
    
# load the finished LDA model from disk
lda = LdaMulticore.load(lda_model_filepath)

Wall time: 344 ms


In [40]:
def explore_topic(topic_number, topn=25):
    """
    accept a user-supplied topic number and
    print out a formatted list of the top terms
    """
        
    print u'{:20} {}'.format(u'term', u'frequency') + u'\n'

    for term, frequency in lda.show_topic(topic_number, topn=25):
        print u'{:20} {:.3f}'.format(term, round(frequency, 3))

In [41]:
explore_topic(topic_number=49)

term                 frequency

review               0.062
yelp                 0.023
know                 0.020
read                 0.020
write                0.014
guy                  0.011
people               0.010
like                 0.010
man                  0.009
let                  0.009
think                0.009
u                    0.008
want                 0.008
's                   0.007
post                 0.007
groupon              0.007
star                 0.006
use                  0.006
mean                 0.005
base                 0.005
yes                  0.005
way                  0.005
question             0.005
update               0.004
girl                 0.004


With the text corpus grouped into clusters, we can explore the words in each cluster to identify a common theme/topic. The list below attempts to identify a common-sense title for each cluster. We can then use this to summarize a review in terms of 3-5 key topics discussed by the reviewer. 

In [42]:
topic_names = {0: u'dessert',
               1: u'table, ordering & service',
               2: u'airport & flying, allergies',
               3: u'buffet & dishes',
               4: u'pizza & italian',
               5: u'bar & drinking',
               6: u'salad & lunch',
               7: u'kids & family',
               8: u'las vegas',
               9: u'fish & asian',
               10: u'seating experience',
               11: u'waffle / cupcake', # waffle + various unrelated terms
               12: u'family members',
               13: u'flavor & food experience',
               14: u'general ambience',
               15: u'night / happy hour / drinks',
               16: u'brunch',
               17: u'mexican',
               18: u'meat',
               19: u'price & paying',
               20: u'wedding / german', # various unrelated words incl single letters
               21: u'various locations', # e.g. north, south, town names, uptown, street etc.
               22: u'cleansiness',
               23: u'mall / cheesecake_factory', # plus various unrelated words and some experience words
               24: u'positive service experience',
               25: u'buying / offers',
               26: u'positive food experience',
               27: u'cheese / wings',
               28: u'french language reviews',
               29: u'positive experience',
               30: u'ordering and service', # geenral and negative service experience
               31: u'time & place',
               32: u'menu & dishes',
               33: u'sushi & fish',
               34: u'german language reviews',
               35: u'positive food experience',
               36: u'timing & waiting',
               37: u'seafood',
               38: u'tea & drinks',
               39: u'thai',
               40: u'staff', 
               41: u'sandwiches',
               42: u'yes / know / eat', # unrelated common words
               43: u'hot_dog, chilli', #hot is the most prominent word, both in teh sense of spicy and hot_dog
               44: u'drinks / coffee / atmosphere',
               45: u'burgers',
               46: u'breakfast food', 
               47: u'price / service / portions',
               48: u'steaks',
               49: u'review / yelp'} # various common urelated words

In [43]:
topic_names_filepath = os.path.join(intermediate_directory, 'topic_names.pkl')

with open(topic_names_filepath, 'w') as f:
    pickle.dump(topic_names, f)

There are a variety of topics related to different types of restaurant, such as **pizza&italian**, **mexican**, **thai**, **sushi&fish**, etc., as well as topics related to the restaurant experience such as seating, service and pricing. 

Beyond these two categories, there are still some topics that are not straightforward to interpret, such as topics 23 and 42.


## Word Vector Embedding with Word2Vec

The goal of *word vector embedding models*, or *word vector models* for short, is to learn dense, numerical vector representations for each term in a corpus vocabulary. If the model is successful, the vectors it learns about each term should encode some information about the *meaning* or *concept* the term represents, and the relationship between it and other terms in the vocabulary. Word vector models are also fully unsupervised &mdash; they learn all of these meanings and relationships solely by analyzing the text of the corpus, without any advance knowledge provided.

Here we will use the _gensim_ [word2vec](https://arxiv.org/pdf/1301.3781v3.pdf) model originally proposed in 2013. The general idea of word2vec is, for a given *focus word*, to use the *context* of the word &mdash; i.e., the other words immediately before and after it &mdash; to provide hints about what the focus word might mean. To do this, word2vec uses a *sliding window* technique, where it considers snippets of text only a few tokens long at a time.

One complete pass sliding the window across all of the corpus text is known as a training *epoch*. We will train the model over multiple passes/epochs in order to optimize the vector representations that are *close* to each other in vector space.

Word2vec has a number of user-defined hyperparameters, including:
- The dimensionality of the vectors - we will use 100-dimensional vector space to allow sufficient complexity at reasonable computing cost
- The width of the sliding window, in tokens - default is used
- The number of training epochs - we will use 5 epochs


In [44]:
from gensim.models import Word2Vec

trigram_sentences = LineSentence(trigram_sentences_filepath)
word2vec_filepath = os.path.join(intermediate_directory, 'word2vec_model_all')

This section trains the word2vec model using the normalized sentences of the tri-gram phrase model in 100-dimensional vector space, using 5 epochs.

In [45]:
if 0 ==1:
    food2vec.train(trigram_sentences, total_examples=food2vec.corpus_count, epochs=food2vec.iter)
    print food2vec.train_count
    food2vec.save(word2vec_filepath)
    
# load the finished model from disk
food2vec = Word2Vec.load(word2vec_filepath)
food2vec.init_sims()

print u'{} training epochs so far.'.format(food2vec.train_count)

2 training epochs so far.


In [46]:
print food2vec.iter
print food2vec.train_count

5
2


In [47]:
print u'{:,} terms in the food2vec vocabulary.'.format(len(food2vec.wv.vocab))

80,588 terms in the food2vec vocabulary.


To look at the word vector model, we can create a pandas DataFrame with the terms as the row labels, and the 100 dimensions of the word vector model as the columns.

In [48]:
# build a list of the terms, integer indices,
# and term counts from the food2vec model vocabulary
ordered_vocab = [(term, voc.index, voc.count)
                 for term, voc in food2vec.wv.vocab.iteritems()]

# sort by the term counts, so the most common terms appear first
ordered_vocab = sorted(ordered_vocab, key=lambda (term, index, count): -count)

# unzip the terms, integer indices, and counts into separate lists
ordered_terms, term_indices, term_counts = zip(*ordered_vocab)

# create a DataFrame with the food2vec vectors as data,
# and the terms as row labels
word_vectors = pd.DataFrame(food2vec.wv.syn0norm[term_indices, :],
                            index=ordered_terms)

word_vectors

0         1         2         3         4   \
-PRON-                -0.075121 -0.067430 -0.127061 -0.009218 -0.022045   
be                     0.012676 -0.047068 -0.190982 -0.033070  0.053919   
the                    0.114931  0.122418 -0.168722  0.071732 -0.059859   
and                    0.114759 -0.138922 -0.144100  0.006292  0.117541   
a                     -0.032536 -0.074984  0.064469 -0.171038  0.020188   
to                     0.040521  0.001736 -0.043932  0.044770 -0.113628   
have                  -0.084267 -0.004397 -0.139282 -0.078336  0.152177   
of                     0.032294 -0.078110 -0.054774  0.063565  0.002737   
not                   -0.052620 -0.045803 -0.088368  0.092387 -0.012385   
for                    0.108099  0.026058  0.099891 -0.059244 -0.024895   
in                     0.041681 -0.015251 -0.094702  0.047722  0.012790   
with                   0.231507  0.061956 -0.046551 -0.069648  0.125986   
that                   0.008432  0.047970 -0.115554  0.048977  0.062427   
but                   -0.153117 -0.052337 -0.165207  0.042091  0.054038   
good                  -0.166842  0.052196 -0.268700  0.081079 -0.000356   
this                  -0.048841  0.099526 -0.119173 -0.009151 -0.039752   
food                  -0.145139 -0.095728 -0.174306  0.003888  0.033305   
on                     0.044580 -0.108453 -0.144697 -0.066781  0.030958   
do                    -0.182227  0.005094 -0.232003  0.144108 -0.072934   
place                 -0.045155 -0.169708 -0.132823 -0.109140 -0.095257   
so                    -0.162822 -0.139266  0.008402 -0.044224  0.139534   
get                    0.050770  0.043213 -0.159441 -0.017592  0.096542   
go                     0.014729  0.050128 -0.226486 -0.143993 -0.051075   
at                     0.049503 -0.119424 -0.033875  0.092584 -0.160535   
order                  0.091938  0.075876 -0.216210  0.031112 -0.025435   
as                    -0.020339 -0.039427 -0.131847  0.084931  0.108922   
great                  0.122305  0.036651 -0.199770  0.035668  0.051312   
very                   0.029132 -0.075060 -0.018703  0.120321  0.095200   
there                 -0.058092 -0.046277 -0.001933  0.072566 -0.031244   
like                   0.051280 -0.042952 -0.046526  0.063202  0.006105   
...                         ...       ...       ...       ...       ...   
active_duty_military   0.075698  0.135177 -0.154208  0.000452 -0.149845   
santana_row            0.071118 -0.063849 -0.171206  0.093245 -0.114105   
seltsame              -0.091998  0.020160 -0.204461 -0.107606 -0.211536   
gastronomisch         -0.044063  0.010626 -0.202993 -0.083327 -0.187213   
mentionner            -0.055281  0.010328 -0.066787 -0.088453 -0.121454   
tlayuda               -0.111358 -0.033820 -0.161888 -0.047562  0.060807   
12.98                 -0.060488  0.009333  0.116836 -0.000846 -0.125824   
1.8                    0.020504 -0.032105  0.030257  0.137854 -0.183497   
la'wan                 0.023135 -0.003220 -0.331206 -0.051811 -0.100299   
copoli                -0.087112  0.207826 -0.171917 -0.090752 -0.225479   
unfiltered_tap        -0.060695 -0.085672 -0.086453  0.040853  0.001239   
das_lokal_liegt       -0.032015 -0.017825 -0.165760 -0.092851 -0.189455   
shredded_iceberg      -0.028121 -0.112043 -0.024992 -0.097683 -0.094441   
panty                 -0.059152  0.101745 -0.085401 -0.043494 -0.187712   
racial_profiling       0.064522  0.093228 -0.106148  0.089736 -0.168122   
softie                 0.077026  0.040195 -0.176717 -0.040961 -0.122685   
2-bite                -0.063560 -0.019889  0.039932 -0.045909  0.009449   
stancatos              0.098584  0.131954 -0.179151 -0.121466 -0.180906   
si_c'était            -0.039697 -0.047400 -0.052664 -0.063865 -0.143274   
rustikale_einrichtung -0.015548  0.002233 -0.195021 -0.099418 -0.219798   
roger_clyne           -0.007507  0.017789 -0.126199 -0.037009 -0.001208   
buerre                 0.037948  0.091147 -0.188829 -0

In [49]:
word_vectors.shape

(80588, 100)

This DataFrame has 80,588 rows &mdash; one for each term in the vocabulary &mdash; and 100 colums. This is the vector representation for each term.

Next, we will save this as a gensim model and it can be reused in another notebook for converting any text into the 100-dimensional vector representation. 

This will provide more _meaning_ to a review, compared to the LDA topic modeling. The word2vec can provide information about what the text is about and if the words have a positive-negative context. The LDA only provides information about the topics (sometimes hard to interpret), but very little information about the positive/negative sentiment of the review. 